fem/fespace.hpp

```diff
@@ -320,12 +400,36 @@ class FiniteElementSpace
    bool Conforming() const { return mesh->Conforming(); }
    bool Nonconforming() const { return mesh->Nonconforming(); }
 
+   /// Sets the order of the i'th finite element.
+   /** By default, all elements are assumed to be of fec->GetOrder(). Once
+       SetElementOrder is called, the space becomes a variable order space. */
+   void SetElementOrder(int i, int p);
+
+   /// Returns the order of the i'th finite element.
+   /** @note This is the 'order' of the FiniteElementCollection. The actual
+       polynomial degree may differ for RT elements, see @a GetElementDegree().*/
+   int GetElementOrder(int i) const;
+
+   /// Return the maximum polynomial order.
+   int GetMaxElementOrder() const
+   { return IsVariableOrder() ? elem_order.Max() : fec->GetOrder(); }
+
+   /// Returns true if the space contains elements of varying polynomial orders.
+   bool IsVariableOrder() const { return elem_order.Size(); }
+
    /// The returned SparseMatrix is owned by the FiniteElementSpace.
    const SparseMatrix *GetConformingProlongation() const;
 
    /// The returned SparseMatrix is owned by the FiniteElementSpace.
    const SparseMatrix *GetConformingRestriction() const;
 
+   /** Return a version of the conforming restriction matrix for variable-order
+       spaces with complex hp interfaces, where some true DOFs are not owned by
+       any elements and need to be interpolated from higher order edge/face
+       variants (see also @a SetRelaxedHpConformity()). */
+   /// The returned SparseMatrix is owned by the FiniteElementSpace.
+   const SparseMatrix *GetHpConformingRestriction() const;
```

Comment:
Is it correct that this method doesn't make sense in most cases?

Reply:
What do you mean in "most cases"? If you don't care about hp interpolation then yes, but if you need to do nodal projections on hp spaces then you need this version of the matrix.

Reply:
Usually a method should make sense in all cases, not only in one case. I would have expected `GetConformingRestriction` to internally check for variable p, and to return the meaningful `SparseMatrix` that is currently returned by `GetHpConformingRestriction` when needed.

Reply:
That was my hope as well, but it turns out in the hp case you need both the standard `R` matrix (for things like `GetEssentialTrueDofs` which does `BooleanMult` where the extra dependencies would screw up the boundary DOFs) *and* the `R_hp` version (e.g. to fix the solution after nodal projection).